lib/accessmonitoring/monitor.go

```diff
@@ -0,0 +1,242 @@
+type EventHandler func(ctx context.Context, event types.Event) error
+
+// Config specifies the access monitor configuration.
+type Config struct {
+	// Logger is the logger for the monitor.
+	Logger *slog.Logger
+
+	// Backend should be a backend.Backend which can be used for obtaining the
+	// lock required to run the service.
+	Backend backend.Backend
+
+	// Events is the event monitor. This interface allows us to monitor access
+	// events.
+	Events types.Events
+}
+
+// CheckAndSetDefaults checks and sets default config values.
+func (c *Config) CheckAndSetDefaults() error {
+	if c.Logger == nil {
+		c.Logger = slog.Default()
+	}
+	if c.Backend == nil {
+		return trace.BadParameter("backend: must be non-nil")
+	}
+	if c.Events == nil {
+		return trace.BadParameter("events: must be non-nil")
+	}
+	return nil
+}
+
+// AccessMonitor is an access monitoring service that monitors access events,
+// then executes the configured set of handlers for each event.
+type AccessMonitor struct {
+	cfg Config
+
+	// ruleHandlers contains the list of access monitoring rule event handlers.
+	ruleHandlers []EventHandler
+	// requestHandlers contains the list access request event handlers.
+	requestHandlers []EventHandler
+}
+
+// NewAccessMonitor returns a new access monitor.
+func NewAccessMonitor(cfg Config) (*AccessMonitor, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &AccessMonitor{
+		cfg: cfg,
+	}, nil
+}
+
+// AddAccessMonitoringRuleHandler appends the access monitoring rule event handler.
+// These handlers are executed whenever a new access monitoring rule event is
+// observed.
+func (s *AccessMonitor) AddAccessMonitoringRuleHandler(handler EventHandler) {
+	s.ruleHandlers = append(s.ruleHandlers, handler)
+}
+
+// AddAccessRequestHandler appends the access request handler.
+// These handlers are executed whenever a new access request event is observed.
+func (s *AccessMonitor) AddAccessRequestHandler(handler EventHandler) {
+	s.requestHandlers = append(s.requestHandlers, handler)
+}
+
+// Run runs the access monitor. Restarts on failure.
+func (s *AccessMonitor) Run(ctx context.Context) error {
+	const retryJitter = 10 * time.Second
+	const retryInterval = 30 * time.Second
+	const lockTTL = time.Minute
+
+	waitWithJitter := retryutils.SeventhJitter(retryJitter)
+	for {
+		err := backend.RunWhileLocked(ctx, backend.RunWhileLockedConfig{
+			LockConfiguration: backend.LockConfiguration{
+				Backend:            s.cfg.Backend,
+				LockNameComponents: []string{componentName},
+				TTL:                lockTTL,
+				RetryInterval:      retryInterval,
+			},
+		}, s.run)
+		if err != nil {
+			s.cfg.Logger.ErrorContext(
+				ctx,
+				"Encountered a fatal error, it will restart after backoff.",
+				"component", componentName,
+				"error", err,
+				"restart_after", waitWithJitter,
+			)
+		}
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-time.After(waitWithJitter):
+		}
+	}
+}
```

Comment:
We had multiple issues with the watcher being stuck to a broken connection and not exiting in kube (I strongly suspect this was caused by a bad grpc connection error detection from our side, and broken kubernetes network sandbox on the user's side).

Because of this we had to introduce a "fail fast" mode when running containerized: https://github.com/gravitational/teleport/pull/30039

I wonder if we should put the env var back in place or if grpc client and watcher improvements will avoid this issue to happen again.

Reply:
Thanks for pointing this out. I'd rather not risk introducing a regression. I'll go ahead and include the "fail fast" mode and include a todo note to investigate if it is still required.